File: instrumentation/core/src/main/java/de/mannodermaus/junit5/condition/EnabledOnManufacturerCondition.java
package de.mannodermaus.junit5.condition;

import android.annotation.TargetApi;
import android.os.Build;
import org.junit.jupiter.api.extension.ConditionEvaluationResult;
import org.junit.jupiter.api.extension.ExecutionCondition;
import org.junit.jupiter.api.extension.ExtensionContext;
import org.junit.platform.commons.util.Preconditions;

import java.util.Arrays;
import java.util.Optional;

import static org.junit.platform.commons.util.AnnotationUtils.findAnnotation;

class EnabledOnManufacturerCondition implements ExecutionCondition {

  private static final ConditionEvaluationResult ENABLED_BY_DEFAULT =
      ConditionEvaluationResult.enabled("@EnabledOnManufacturer is not present");

  static ConditionEvaluationResult enabled() {
    return ConditionEvaluationResult.enabled("Enabled on Manufacturer: " + Build.MANUFACTURER);
  }

  static ConditionEvaluationResult disabled() {
    return ConditionEvaluationResult.disabled("Disabled on Manufacturer: " + Build.MANUFACTURER);
  }

  @TargetApi(24)
  @Override
  public ConditionEvaluationResult evaluateExecutionCondition(ExtensionContext context) {
    Optional<EnabledOnManufacturer> optional = findAnnotation(context.getElement(), EnabledOnManufacturer.class);

    if (optional.isPresent()) {
      EnabledOnManufacturer annotation = optional.get();
      String[] patterns = annotation.value();
      boolean ignoreCase = annotation.ignoreCase();
      Preconditions.condition(patterns.length > 0, "You must declare at least one Manufacturer in @EnabledOnManufacturer");

      return Arrays.stream(patterns).anyMatch(value -> matchesCurrentManufacturer(value, ignoreCase))
          ? enabled()
          : disabled();
    }

    return ENABLED_BY_DEFAULT;
  }

  private boolean matchesCurrentManufacturer(String value, boolean ignoreCase) {
    if (ignoreCase) {
      return Build.MANUFACTURER.equalsIgnoreCase(value);
    } else {
      return Build.MANUFACTURER.equals(value);
    }
  }
}

File: instrumentation/core/src/main/java/de/mannodermaus/junit5/internal/EnabledOnManufacturerCondition.kt
package de.mannodermaus.junit5.internal

import android.annotation.TargetApi
import android.os.Build
import de.mannodermaus.junit5.condition.EnabledOnManufacturer
import org.junit.jupiter.api.extension.ConditionEvaluationResult
import org.junit.jupiter.api.extension.ExecutionCondition
import org.junit.jupiter.api.extension.ExtensionContext
import org.junit.platform.commons.util.AnnotationUtils.findAnnotation
import org.junit.platform.commons.util.Preconditions

internal class EnabledOnManufacturerCondition : ExecutionCondition {

    companion object {
        private val ENABLED_BY_DEFAULT =
            ConditionEvaluationResult.enabled("@EnabledOnManufacturer is not present")

        fun enabled(): ConditionEvaluationResult {
            return ConditionEvaluationResult.enabled("Enabled on Manufacturer: " + Build.MANUFACTURER)
        }

        fun disabled(): ConditionEvaluationResult {
            return ConditionEvaluationResult.disabled("Disabled on Manufacturer: " + Build.MANUFACTURER)
        }
    }

    @TargetApi(24)
    override fun evaluateExecutionCondition(context: ExtensionContext): ConditionEvaluationResult {
        val optional = findAnnotation(context.element, EnabledOnManufacturer::class.java)

        if (optional.isPresent) {
            val annotation = optional.get()
            val patterns = annotation.value
            val ignoreCase = annotation.ignoreCase

            Preconditions.condition(
                patterns.isNotEmpty(),
                "You must declare at least one Manufacturer in @EnabledOnManufacturer"
            )

            return if (patterns.any { Build.MANUFACTURER.equals(it, ignoreCase = ignoreCase) }) {
                enabled()
            } else {
                disabled()
            }
        }

        return ENABLED_BY_DEFAULT
    }
}
